nucypher/blockchain/economics.py

```diff
@@ -14,60 +14,166 @@
+    maximum_allowed_locked = int(4e6) * nunits_per_token
+
+    # Supply
+    __default_initial_supply = Decimal(int(1e9) * nunits_per_token)
+    __default_initial_inflation = Decimal(1.0)
+    __default_token_halving = Decimal(2.0)     # years
+    __default_reward_saturation = Decimal(1.0)  # years
+    __default_small_stake_multiplier = Decimal(0.5)
+
+    def __init__(self,
+                 initial_supply: Decimal = __default_initial_supply,
+                 initial_inflation: Decimal = __default_initial_inflation,
+                 halving_delay: Decimal = __default_token_halving,
+                 reward_saturation: Decimal = __default_reward_saturation,
+                 small_stake_multiplier: Decimal = __default_small_stake_multiplier
+                 ):
         """
-        Calculate parameters to use in token issuer from high-level
-        human-understandable parameters
-
-        :param initial_supply: coins at t=0,
-        :param initial_inflation: inflation in day 1 expressed in units of year**-1
-            e.g. 1/365 of all coins in day 1 is 1 year**-1
-        :param T_half: time for inflation halving in years
-        :param T_sat: "saturation" time - if staking is longer than T_sat, the
-            reward doesn't get higher
-        :param small_staker_multiplier: what fraction of max reward rate do we
-            pay to those who are about to unlock coins
+        :param initial_supply: Tokens at t=0
+        :param initial_inflation: Inflation on day 1 expressed in units of year**-1
+        :param halving_delay: Time for inflation halving in years
+        :param reward_saturation: "saturation" time - if staking is longer than T_sat, the reward doesn't get any higher
+        :param small_stake_multiplier: Fraction of maximum reward rate paid to those who are about to unlock tokens
         """
 
-        # Saving input in case we'll need it
-        self.input_params = dict(
-                initial_supply=initial_supply,
-                initial_inflation=initial_inflation,
-                T_half=T_half,
-                T_sat=T_sat,
-                small_staker_multiplier=small_staker_multiplier)
+        self.initial_supply = initial_supply
+        self.initial_inflation = initial_inflation
+        self.token_halving = halving_delay
+        self.token_saturation = reward_saturation
+        self.small_stake_multiplier = small_stake_multiplier
+        self.log_scale = Decimal(log(2))
+
+        #
+        # Calculate
+        #
+
+        # ERC20 Token parameter
+        self.__total_supply = int(initial_supply * (1 + initial_inflation * halving_delay / self.log_scale))
+
+        # Remaining / Reward Supply - Escrow Parameter
+        self.__reward_supply = int(self.total_supply - self.initial_supply)
+
+        # k2 - Escrow parameter
+        self.__staking_coefficient = 365 ** 2 * reward_saturation * halving_delay / self.log_scale / (1 - small_stake_multiplier)
```

Comment:
Wow, what's going on here with `halving_delay / self.log_scale / (1 - small_stake_multiplier)`? It's an odd expression. Note that the '/' operator is not associative:
```python
>>> 1/2/3
0.16666666666666666
>>> (1/2)/3
0.16666666666666666
>>> 1/(2/3)
1.5
```
This expression needs parenthesis in order to remove any potential ambiguity.

Reply:
Related to your above comment we can check this against the reference.

Reply:
@michwill - Perhaps you can verify the order of ops against the reference?

Reply:
Well, I just don't like when braces go together like `))`, so instead of `a / (b * c)` I wrote `a / b / c`. Double-checking though...

Reply:
This looks correct to me.
Equivalently, you could write `365 ** 2 * reward_saturation * halving_delay / (self.log_scale * (1 - small_stake_multiplier))` if you feel it to be more readable (I don't)

Reply:
@michwill what about `365 ** 2 * (reward_saturation * halving_delay / LOG2) / (1 - small_stake_multiplier)`? That way there are no two braces together like `))`. 
On the other hand, this made me learn that the operator `/` is implicitly left-associative, so I'm also fine with the original expression.

Reply:
Could equally be `365 ** 2 / LOG2 * ....`. They are all equally readable to me though